amads/time/meter/attractor_tempos.py

```diff
@@ -0,0 +1,200 @@
+"""
+What does it mean for music to be "fast" or "slow"?
+Certainly BPM is not enough.
+The "Attractor tempos" theory (Gotham 2015, [1]) proposes
+ a definition of "fast"/"slow" relative to neutral, central, moderate tempos
+ and a definition of those moderate ("Attractor") tempos that accounts for the role of the metrical structure.
+In short, it provides a model for optimizing the salience of metrical structures.
+
+[1] Gotham, M. (2015). Attractor tempos for metrical structures. Journal of Mathematics and Music, 9(1), 23–44.
+https://doi.org/10.1080/17459737.2014.980343
+"""
+
+__author__ = "Mark Gotham"
+
+from typing import Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+# ------------------------------------------------------------------------------
+
+
+class MetricalSalience:
+    """
+    Organise array representations of metrical structre and derived salience values.
+
+    Parameters
+    ----------
+    symbolic_pulse_length_array: A NumPy array representing the symbolic pulse lengths.
+
+    Examples
+    --------
+
+    >>> from amads.time.meter import PulseLengths
+    >>> pl = [4, 2, 1, 0.5]
+    >>> pls = PulseLengths(pulse_lengths=pl, cycle_length=4)
+    >>> arr = pls.to_array()
+    >>> arr
+    array([[4. , 0. , 0. , 0. , 0. , 0. , 0. , 0. ],
+           [2. , 0. , 0. , 0. , 2. , 0. , 0. , 0. ],
+           [1. , 0. , 1. , 0. , 1. , 0. , 1. , 0. ],
+           [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]])
+
+    >>> ms = MetricalSalience(symbolic_pulse_length_array=arr)
+    >>> ms.calculate_absolute_pulse_lengths(quarter_bpm=120)
+    >>> ms.absolute_pulse_length_array
+    array([[2.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ],
+           [1.  , 0.  , 0.  , 0.  , 1.  , 0.  , 0.  , 0.  ],
+           [0.5 , 0.  , 0.5 , 0.  , 0.5 , 0.  , 0.5 , 0.  ],
+           [0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25]])
+
+    >>> ms.calculate_salience_values()
+    >>> ms.calculate_cumulative_salience_values()
+    >>> ms.cumulative_salience_values
+    array([2.39342011, 0.44793176, 1.4136999 , 0.44793176, 2.17446773,
+           0.44793176, 1.4136999 , 0.44793176])
+
+    """
+
+    def __init__(
+        self,
+        symbolic_pulse_length_array: Optional[np.array] = None,
+    ):
+        self.symbolic_pulse_length_array = symbolic_pulse_length_array
+        self.pulse_lengths = self.symbolic_pulse_length_array[:, 0]
```

Comment:
The type hint says `None` is a valid value for `symbolic_pulse_length_array`, but `symbolic_pulse_length_array[:, 0]` will fail in such a case

Reply:
True! Fixed.

Done.